Web Inspector: Worker Info not Visible In Most Timeline Profile Views like JavaScript & Events

dcrousso opened the pull request:
#### 783a08ad6950cdf316d3f4ab3807797f09f84dc0
<pre>
Web Inspector: Worker Info not Visible In Most Timeline Profile Views like JavaScript &amp; Events
<a href="https://bugs.webkit.org/show_bug.cgi?id=263334">https://bugs.webkit.org/show_bug.cgi?id=263334</a>
<a href="https://rdar.apple.com/117491838">rdar://117491838</a>

Reviewed by BJ Burg.

This will allow developers to investigate performance issues in `Worker` using the Timelines Tab just like they can already do for other page content (i.e. main thread).

* Source/JavaScriptCore/inspector/protocol/Timeline.json:
* Source/WebCore/inspector/WorkerInspectorController.cpp:
(WebCore::WorkerInspectorController::createLazyAgents):
Expose `Timeline` in `Worker`.

* Source/WebCore/inspector/agents/InspectorTimelineAgent.h:
(WebCore::InspectorTimelineAgent::shouldStartHeapInstrument const):
(WebCore::InspectorTimelineAgent::instruments const):
* Source/WebCore/inspector/agents/InspectorTimelineAgent.cpp:
(WebCore::InspectorTimelineAgent::InspectorTimelineAgent):
(WebCore::InspectorTimelineAgent::enable):
(WebCore::InspectorTimelineAgent::disable):
(WebCore::InspectorTimelineAgent::start):
(WebCore::InspectorTimelineAgent::stop):
(WebCore::InspectorTimelineAgent::enabled const):
(WebCore::InspectorTimelineAgent::internalEnable):
(WebCore::InspectorTimelineAgent::internalDisable):
(WebCore::InspectorTimelineAgent::tracking const):
(WebCore::InspectorTimelineAgent::internalStart):
(WebCore::InspectorTimelineAgent::internalStop):
(WebCore::InspectorTimelineAgent::autoCaptureStarted const):
(WebCore::InspectorTimelineAgent::startFromConsole):
(WebCore::InspectorTimelineAgent::startProgrammaticCapture):
(WebCore::InspectorTimelineAgent::stopProgrammaticCapture):
(WebCore::InspectorTimelineAgent::toggleHeapInstrument):
(WebCore::InspectorTimelineAgent::toggleTimelineInstrument):
(WebCore::InspectorTimelineAgent::lastRecordEntry):
(WebCore::currentRunLoop): Deleted.
(WebCore::InspectorTimelineAgent::setAutoCaptureEnabled): Deleted.
(WebCore::InspectorTimelineAgent::didInvalidateLayout): Deleted.
(WebCore::InspectorTimelineAgent::willLayout): Deleted.
(WebCore::InspectorTimelineAgent::didLayout): Deleted.
(WebCore::InspectorTimelineAgent::didScheduleStyleRecalculation): Deleted.
(WebCore::InspectorTimelineAgent::willRecalculateStyle): Deleted.
(WebCore::InspectorTimelineAgent::didRecalculateStyle): Deleted.
(WebCore::InspectorTimelineAgent::willComposite): Deleted.
(WebCore::InspectorTimelineAgent::didComposite): Deleted.
(WebCore::InspectorTimelineAgent::willPaint): Deleted.
(WebCore::InspectorTimelineAgent::didPaint): Deleted.
(WebCore::InspectorTimelineAgent::mainFrameStartedLoading): Deleted.
(WebCore::InspectorTimelineAgent::mainFrameNavigated): Deleted.
(WebCore::InspectorTimelineAgent::captureScreenshot): Deleted.
(WebCore::InspectorTimelineAgent::didCompleteRenderingFrame): Deleted.
* Source/WebCore/inspector/agents/page/PageTimelineAgent.h: Added.
* Source/WebCore/inspector/agents/page/PageTimelineAgent.cpp: Added.
(WebCore::currentRunLoop):
(WebCore::PageTimelineAgent::PageTimelineAgent):
(WebCore::PageTimelineAgent::enabled const):
(WebCore::PageTimelineAgent::internalEnable):
(WebCore::PageTimelineAgent::internalDisable):
(WebCore::PageTimelineAgent::tracking const):
(WebCore::PageTimelineAgent::internalStart):
(WebCore::PageTimelineAgent::internalStop):
(WebCore::PageTimelineAgent::setAutoCaptureEnabled):
(WebCore::PageTimelineAgent::didInvalidateLayout):
(WebCore::PageTimelineAgent::willLayout):
(WebCore::PageTimelineAgent::didLayout):
(WebCore::PageTimelineAgent::didScheduleStyleRecalculation):
(WebCore::PageTimelineAgent::willRecalculateStyle):
(WebCore::PageTimelineAgent::didRecalculateStyle):
(WebCore::PageTimelineAgent::willComposite):
(WebCore::PageTimelineAgent::didComposite):
(WebCore::PageTimelineAgent::willPaint):
(WebCore::PageTimelineAgent::didPaint):
(WebCore::PageTimelineAgent::mainFrameStartedLoading):
(WebCore::PageTimelineAgent::mainFrameNavigated):
(WebCore::PageTimelineAgent::didCompleteRenderingFrame):
(WebCore::PageTimelineAgent::shouldStartHeapInstrument const):
(WebCore::PageTimelineAgent::captureScreenshot):
* Source/WebCore/inspector/agents/worker/WorkerTimelineAgent.h: Added.
* Source/WebCore/inspector/agents/worker/WorkerTimelineAgent.cpp: Added.
(WebCore::WorkerTimelineAgent::WorkerTimelineAgent):
(WebCore::WorkerTimelineAgent::setAutoCaptureEnabled):
* Source/WebCore/inspector/InstrumentingAgents.h:
* Source/WebCore/inspector/InspectorController.cpp:
(WebCore::InspectorController::createLazyAgents):
* Source/WebCore/inspector/InspectorInstrumentation.cpp:
(WebCore::InspectorInstrumentation::didInvalidateLayoutImpl):
(WebCore::InspectorInstrumentation::willLayoutImpl):
(WebCore::InspectorInstrumentation::didLayoutImpl):
(WebCore::InspectorInstrumentation::willCompositeImpl):
(WebCore::InspectorInstrumentation::didCompositeImpl):
(WebCore::InspectorInstrumentation::willPaintImpl):
(WebCore::InspectorInstrumentation::didPaintImpl):
(WebCore::InspectorInstrumentation::willRecalculateStyleImpl):
(WebCore::InspectorInstrumentation::didRecalculateStyleImpl):
(WebCore::InspectorInstrumentation::didScheduleStyleRecalculationImpl):
(WebCore::InspectorInstrumentation::didCommitLoadImpl):
(WebCore::InspectorInstrumentation::frameStartedLoadingImpl):
(WebCore::InspectorInstrumentation::didCompleteRenderingFrameImpl):
Split the existing `InspectorTimelineAgent` into a `PageTimelineAgent` and `WorkerTimelineAgent`.
This is mainly done for clarity as certain concepts (e.g. compositing, navigation, screenshots, etc.) don&apos;t make sense in a `Worker`.

* Source/WebCore/workers/WorkerOrWorkletScriptController.h:
* Source/WebCore/workers/WorkerOrWorkletScriptController.cpp:
(WebCore::WorkerOrWorkletScriptController::evaluate):
(WebCore::WorkerOrWorkletScriptController::evaluateModule):
* Source/WebCore/bindings/js/ScriptModuleLoader.cpp:
(WebCore::ScriptModuleLoader::evaluate):
* Source/WebCore/inspector/InspectorInstrumentation.h:
(WebCore::InspectorInstrumentation::willEvaluateScript):
(WebCore::InspectorInstrumentation::didEvaluateScript):
Add `InspectorInstrumentation` hooks before and after evaluating JS in `Worker`.

* Source/JavaScriptCore/inspector/agents/InspectorHeapAgent.h:
* Source/WebCore/inspector/agents/WebHeapAgent.h:
* Source/WebCore/inspector/agents/WebHeapAgent.cpp:
(WebCore::WebHeapAgent::didCreateFrontendAndBackend):
(WebCore::WebHeapAgent::willDestroyFrontendAndBackend):
Also expose `WebHeapAgent` via `InstrumentingAgents` instead of only just `PageHeapAgent`.
This is needed by `InspectorTimelineAgent` to start capturing heap snapshots during a timeline recording.

* Source/WebInspectorUI/UserInterface/Models/Instrument.js:
(WI.Instrument.startLegacyTimelineAgent):
(WI.Instrument.stopLegacyTimelineAgent):
Iterate over all `WI.Target` when starting a timeline recording now that `Timeline` also exists in `Worker`.

* Source/WebInspectorUI/UserInterface/Protocol/TimelineObserver.js:
(WI.TimelineObserver.prototype.eventRecorded):
* Source/WebInspectorUI/UserInterface/Controllers/HeapManager.js:
(WI.HeapManager.prototype.garbageCollected):
Pass along the `WI.Target`.

* Source/WebInspectorUI/UserInterface/Controllers/TimelineManager.js:
(WI.TimelineManager):
(WI.TimelineManager.prototype.initializeTarget):
(WI.TimelineManager.prototype.capturingStarted):
(WI.TimelineManager.prototype.eventRecorded):
(WI.TimelineManager.prototype.scriptProfilerTrackingUpdated):
(WI.TimelineManager.prototype.scriptProfilerTrackingCompleted):
(WI.TimelineManager.prototype._processRecord):
(WI.TimelineManager.prototype._processEvent):
(WI.TimelineManager.prototype._stackTraceFromPayload):
(WI.TimelineManager.prototype._garbageCollected):
Have a separate array of `WI.ScriptTimelineRecord` for each `WI.Target`.

* Source/WebInspectorUI/UserInterface/Models/TimelineRecording.js:
(WI.TimelineRecording):
(WI.TimelineRecording.async import):
(WI.TimelineRecording.prototype.get targets): Added.
(WI.TimelineRecording.prototype.callingContextTree): Added.
(WI.TimelineRecording.prototype.reset):
(WI.TimelineRecording.prototype.updateCallingContextTrees): Renamed from `initializeCallingContextTrees`.
(WI.TimelineRecording.prototype.get topDownCallingContextTree): Deleted.
(WI.TimelineRecording.prototype.get bottomUpCallingContextTree): Deleted.
(WI.TimelineRecording.prototype.get topFunctionsTopDownCallingContextTree): Deleted.
(WI.TimelineRecording.prototype.get topFunctionsBottomUpCallingContextTree): Deleted.
Have separate `WI.CallingContextTree` for each `WI.Target` (and for each `WI.CallingContextTree.Type`).
Callers must `updateCallingContextTrees` with a valid `WI.Target` before attempting to fetch a `WI.CallingContextTree` for that `WI.Target` (and `WI.CallingContextTree.Type`).

* Source/WebInspectorUI/UserInterface/Models/ScriptTimelineRecord.js:
(WI.ScriptTimelineRecord.async fromJSON):
(WI.ScriptTimelineRecord.prototype.toJSON):
(WI.ScriptTimelineRecord.prototype.get target): Added.
(WI.ScriptTimelineRecord.prototype._initializeProfileFromPayload.profileNodeFromPayload):
* Source/WebInspectorUI/UserInterface/Models/CallingContextTree.js:
(WI.CallingContextTree):
(WI.CallingContextTree.prototype.get target): Added.
(WI.CallingContextTree.prototype.reset): Deleted.
(WI.CallingContextTree.__test_makeTreeFromProtocolMessageObject): Deleted.
(WI.CallingContextTree.prototype.__test_matchesStackTrace): Deleted.
(WI.CallingContextTree.prototype.__test_buildLeafLinkedLists): Deleted.
* Source/WebInspectorUI/UserInterface/Views/ProfileDataGridTree.js:
(WI.ProfileDataGridTree.prototype.get target): Added.
* Source/WebInspectorUI/UserInterface/Views/ProfileDataGridNode.js:
(WI.ProfileDataGridNode.prototype.iconClassName):
(WI.ProfileDataGridNode.prototype.filterableDataForColumn):
(WI.ProfileDataGridNode.prototype._displayContent):
Save and use the relevant `WI.Target`.

* Source/WebInspectorUI/UserInterface/Views/ScriptDetailsTimelineView.js:
(WI.ScriptDetailsTimelineView):
(WI.ScriptDetailsTimelineView.prototype.closed):
(WI.ScriptDetailsTimelineView.prototype._processPendingRecords):
* Source/WebInspectorUI/UserInterface/Views/ScriptTimelineDataGridNode.js:
(WI.ScriptTimelineDataGridNode.prototype.get data):
Show the name of the `WI.Target` (a.k.a. &quot;Context&quot;) when there&apos;s more than one.

* Source/WebInspectorUI/UserInterface/Views/ScriptProfileTimelineView.js:
(WI.ScriptProfileTimelineView):
(WI.ScriptProfileTimelineView.prototype.closed):
(WI.ScriptProfileTimelineView.prototype.get navigationItems):
(WI.ScriptProfileTimelineView.prototype._callingContextTreeForOrientation):
(WI.ScriptProfileTimelineView.prototype._updateTargetNavigationItemDisplay):
(WI.ScriptProfileTimelineView.prototype._displayNameForTarget):
(WI.ScriptProfileTimelineView.prototype._profileOrientationButtonClicked):
(WI.ScriptProfileTimelineView.prototype._topFunctionsButtonClicked):
(WI.ScriptProfileTimelineView.prototype._createProfileView):
(WI.ScriptProfileTimelineView.prototype._showProfileView):
(WI.ScriptProfileTimelineView.prototype._handleRecordingTargetAdded):
(WI.ScriptProfileTimelineView.prototype._populateTargetNavigationItemContextMenu):
(WI.ScriptProfileTimelineView.prototype._showProfileViewForOrientation): Deleted.
* Source/WebInspectorUI/UserInterface/Views/ScriptProfileTimelineView.css: Added.
(.content-view.timeline-recording .navigation-bar &gt; .script-profile-target &gt; .selector-arrows):
Create a dropdown for all `WI.Target` when there&apos;s more than one to control which `WI.CallingContextTree` is shown.
We probably don&apos;t want to merge stack trace data from different `WI.Target` as it could be misleading if a function (or at least it&apos;s name and source code location) is used in multiple `WI.Target`.

* Source/WebInspectorUI/UserInterface/Views/ClusterContentView.js:
(WI.ClusterContentView):
(WI.ClusterContentView.prototype.closed):
(WI.ClusterContentView.prototype._contentViewNavigationItemsDidChange):
This appears to be the first time that a `WI.ContentView` inside a `WI.ClusterContentView` ever dynamically changes the result of `get navigationItems`, so dispatch a `WI.ContentView.Event.NavigationItemsDidChange` to the parent `WI.ContentBrowser` so that it&apos;s updated accordingly.

* Source/WebInspectorUI/UserInterface/Views/DataGrid.js:
(WI.DataGrid.prototype.insertColumn): All cells need to be refreshed when a new column is inserted.
(WI.DataGrid.prototype.setColumnVisible): When a new column is first inserted it doesn&apos;t have a previous `hidden` state, so always make sure to set one.
(WI.DataGrid.prototype._contextMenuInHeader): Columns should be listed by ordinal instead of insertion order.

* Source/WebCore/page/PerformanceUserTiming.cpp:
(WebCore::PerformanceUserTiming::mark):
* Source/WebInspectorUI/UserInterface/Models/ScriptInstrument.js:
(WI.ScriptInstrument.prototype.stopInstrumentation):
Drive-by: remove unused variable.

* Source/WebInspectorUI/UserInterface/Views/CPUTimelineView.js:
(WI.CPUTimelineView.prototype.closed):
* Source/WebInspectorUI/UserInterface/Views/HeapAllocationsTimelineView.js:
(WI.HeapAllocationsTimelineView.prototype.closed):
* Source/WebInspectorUI/UserInterface/Views/LayoutTimelineView.js:
(WI.LayoutTimelineView.prototype.closed):
* Source/WebInspectorUI/UserInterface/Views/MemoryTimelineView.js:
(WI.MemoryTimelineView.prototype.closed):
* Source/WebInspectorUI/UserInterface/Views/NetworkTimelineView.js:
(WI.NetworkTimelineView.prototype.closed):
* Source/WebInspectorUI/UserInterface/Views/OverviewTimelineView.js:
(WI.OverviewTimelineView.prototype.closed):
* Source/WebInspectorUI/UserInterface/Views/RenderingFrameTimelineView.js:
(WI.RenderingFrameTimelineView.prototype.closed):
Drive-by: add missing `super` call.

* Source/WebInspectorUI/UserInterface/Views/SettingsTabContentView.js:
(WI.SettingsTabContentView.prototype._createExperimentalSettingsView):
Drive-by: add missing separator.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebInspectorUI/UserInterface/Main.html:
* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:

* Source/WebInspectorUI/UserInterface/Models/CallingContextTreeNode.js:
(WI.CallingContextTreeNode.prototype.__test_buildLeafLinkedLists): Deleted.
* LayoutTests/inspector/sampling-profiler/basic.html:
* LayoutTests/inspector/sampling-profiler/call-frame-with-dom-functions.html:
* LayoutTests/inspector/sampling-profiler/eval-source-url.html:
* LayoutTests/inspector/sampling-profiler/expression-location-info.html:
* LayoutTests/inspector/sampling-profiler/expression-location-info-expected.txt:
* LayoutTests/inspector/sampling-profiler/many-call-frames.html:
* LayoutTests/inspector/sampling-profiler/named-function-expression.html:
* LayoutTests/inspector/sampling-profiler/resources/calling-context-tree.js: Added.
Move test functions to test files so they&apos;re not shipped in production builds.

Canonical link: <a href="https://commits.webkit.org/290430@main">https://commits.webkit.org/290430@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bac739300fb881cc502bec577561e20e1faa3ce9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/90010 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/9539 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/44910 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/95010 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/40783 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/92062 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/9926 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/17825 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/69292 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26896 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/93011 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/7594 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/81656 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49650 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/7322 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/36016 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/39917 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/82813 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/77657 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/37083 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/96836 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/88788 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/17198 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/12615 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/78294 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/17454 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/77480 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/77497 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19135 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21946 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/20534 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/10392 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/17208 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/22533 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/111279 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/16949 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/26640 "Found 9 new JSC stress test failures: microbenchmarks/memcpy-wasm-medium.js.mini-mode, wasm.yaml/wasm/stress/b3-signed-extend-16-to-64.js.wasm-eager-jettison, wasm.yaml/wasm/stress/b3-signed-extend-16-to-64.js.wasm-slow-memory, wasm.yaml/wasm/stress/live-funcref-across-loop-tier-up.js.wasm-eager, wasm.yaml/wasm/stress/live-funcref-across-loop-tier-up.js.wasm-slow-memory, wasm.yaml/wasm/stress/repro_1289.js.default-wasm, wasm.yaml/wasm/stress/repro_1289.js.wasm-collect-continuously, wasm.yaml/wasm/stress/repro_1289.js.wasm-eager, wasm.yaml/wasm/stress/repro_1289.js.wasm-eager-jettison (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/20401 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/18732 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

burg commented on LayoutTests/inspector/sampling-profiler/basic.html:
```diff
@@ -29,12 +30,12 @@
         test(resolve, reject) {
             InspectorProtocol.awaitEvent({event: "ScriptProfiler.trackingComplete"}).then((messageObject) => {
-                let tree = WI.CallingContextTree.__test_makeTreeFromProtocolMessageObject(messageObject);
+                let tree = ProtocolTest.CallingContextTree.createFromProtocolMessageObject(messageObject);
```
OK

burg commented on Source/WebInspectorUI/UserInterface/Models/TimelineRecording.js:
```diff
@@ -76,6 +73,8 @@ WI.TimelineRecording = class TimelineRecording extends WI.Object
 
     static async import(identifier, json, displayName)
     {
+        const target = WI.assumingMainTarget();
```
I'd like to start adding FIXMEs to `WI.assumingMainTarget()`, can you file a followup bugzilla bug about multi-target Timeline import/export? (Should also add the same FIXME in the `toJSON()` method above)

dcrousso replied:
so actually the original idea behind `WI.assumingMainTarget()` was that it's implicitly a FIXME since we should never be "assuming" anything 😛

eventually we should be able to entirely remove `WI.assumingMainTarget()` as everything will either support multiple targets or explicitly only affect the `WI.mainTarget`

burg commented on Source/WebInspectorUI/UserInterface/Models/TimelineRecording.js:
```diff
@@ -154,10 +155,16 @@ WI.TimelineRecording = class TimelineRecording extends WI.Object
+        return Array.from(this._callingContextTreesForTarget.keys());
+    }
+
+    callingContextTree(target, type)
```
Since we return multiple trees in a Map, I was expecting `callingContextTrees` or `callingContextTreesForTarget` as the method name. Is there a reason to use the singular?

dcrousso replied:
this should only return a single `WI.CallingContextTree`

burg commented on Source/WebInspectorUI/UserInterface/Models/TimelineRecording.js:
```diff
@@ -427,17 +431,32 @@ WI.TimelineRecording = class TimelineRecording extends WI.Object
     {
-        this._exportDataSampleStackTraces.pushAll(stackTraces);
-        this._exportDataSampleDurations.pushAll(sampleDurations);
+        // FIXME: support exporting and importing data from worker targets
```
Please add link as discussed above

burg commented on Source/WebCore/inspector/agents/InspectorTimelineAgent.h:
```diff
@@ -114,52 +112,41 @@ class InspectorTimelineAgent final : public InspectorAgentBase , public Inspecto
-    void mainFrameStartedLoading();
-    void mainFrameNavigated();
-    void didCompleteRenderingFrame();
+    void time(const String& lagel);
```
Nit: label

burg commented on Source/WebCore/inspector/agents/page/PageTimelineAgent.cpp:
```diff
@@ -0,0 +1,338 @@
+/*
+ * Copyright (C) 2025 Devin Rousso <webkit@devinrousso.com>. All rights reserved.
```
It is my preference that we copy the bylines from the file that is being split up.

burg commented on Source/WebCore/inspector/agents/page/PageTimelineAgent.h:
```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (C) 2025 Devin Rousso <webkit@devinrousso.com>. All rights reserved.
```
Ditto (applies throughout)

burg commented on Source/WebInspectorUI/UserInterface/Views/ScriptProfileTimelineView.js:
```diff
@@ -201,6 +254,53 @@ WI.ScriptProfileTimelineView = class ScriptProfileTimelineView extends WI.Timeli
+            (target) => target.type === WI.TargetType.Page,
+            (target) => target.type === WI.TargetType.Worker,
+        ];
+        let sortedTargets = this._recording.targets.sort((a, b) => {
```
Neat.

dcrousso committed:
Web Inspector: Worker Info not Visible In Most Timeline Profile Views like JavaScript & Events
https://bugs.webkit.org/show_bug.cgi?id=263334
rdar://117491838

Reviewed by BJ Burg.

This will allow developers to investigate performance issues in `Worker` using the Timelines Tab just like they can already do for other page content (i.e. main thread).

* Source/JavaScriptCore/inspector/protocol/Timeline.json:
* Source/WebCore/inspector/WorkerInspectorController.cpp:
(WebCore::WorkerInspectorController::createLazyAgents):
Expose `Timeline` in `Worker`.

* Source/WebCore/inspector/agents/InspectorTimelineAgent.h:
(WebCore::InspectorTimelineAgent::shouldStartHeapInstrument const):
(WebCore::InspectorTimelineAgent::instruments const):
* Source/WebCore/inspector/agents/InspectorTimelineAgent.cpp:
(WebCore::InspectorTimelineAgent::InspectorTimelineAgent):
(WebCore::InspectorTimelineAgent::enable):
(WebCore::InspectorTimelineAgent::disable):
(WebCore::InspectorTimelineAgent::start):
(WebCore::InspectorTimelineAgent::stop):
(WebCore::InspectorTimelineAgent::enabled const):
(WebCore::InspectorTimelineAgent::internalEnable):
(WebCore::InspectorTimelineAgent::internalDisable):
(WebCore::InspectorTimelineAgent::tracking const):
(WebCore::InspectorTimelineAgent::internalStart):
(WebCore::InspectorTimelineAgent::internalStop):
(WebCore::InspectorTimelineAgent::autoCaptureStarted const):
(WebCore::InspectorTimelineAgent::startFromConsole):
(WebCore::InspectorTimelineAgent::startProgrammaticCapture):
(WebCore::InspectorTimelineAgent::stopProgrammaticCapture):
(WebCore::InspectorTimelineAgent::toggleHeapInstrument):
(WebCore::InspectorTimelineAgent::toggleTimelineInstrument):
(WebCore::InspectorTimelineAgent::lastRecordEntry):
(WebCore::currentRunLoop): Deleted.
(WebCore::InspectorTimelineAgent::setAutoCaptureEnabled): Deleted.
(WebCore::InspectorTimelineAgent::didInvalidateLayout): Deleted.
(WebCore::InspectorTimelineAgent::willLayout): Deleted.
(WebCore::InspectorTimelineAgent::didLayout): Deleted.
(WebCore::InspectorTimelineAgent::didScheduleStyleRecalculation): Deleted.
(WebCore::InspectorTimelineAgent::willRecalculateStyle): Deleted.
(WebCore::InspectorTimelineAgent::didRecalculateStyle): Deleted.
(WebCore::InspectorTimelineAgent::willComposite): Deleted.
(WebCore::InspectorTimelineAgent::didComposite): Deleted.
(WebCore::InspectorTimelineAgent::willPaint): Deleted.
(WebCore::InspectorTimelineAgent::didPaint): Deleted.
(WebCore::InspectorTimelineAgent::mainFrameStartedLoading): Deleted.
(WebCore::InspectorTimelineAgent::mainFrameNavigated): Deleted.
(WebCore::InspectorTimelineAgent::captureScreenshot): Deleted.
(WebCore::InspectorTimelineAgent::didCompleteRenderingFrame): Deleted.
* Source/WebCore/inspector/agents/page/PageTimelineAgent.h: Added.
* Source/WebCore/inspector/agents/page/PageTimelineAgent.cpp: Added.
(WebCore::currentRunLoop):
(WebCore::PageTimelineAgent::PageTimelineAgent):
(WebCore::PageTimelineAgent::enabled const):
(WebCore::PageTimelineAgent::internalEnable):
(WebCore::PageTimelineAgent::internalDisable):
(WebCore::PageTimelineAgent::tracking const):
(WebCore::PageTimelineAgent::internalStart):
(WebCore::PageTimelineAgent::internalStop):
(WebCore::PageTimelineAgent::setAutoCaptureEnabled):
(WebCore::PageTimelineAgent::didInvalidateLayout):
(WebCore::PageTimelineAgent::willLayout):
(WebCore::PageTimelineAgent::didLayout):
(WebCore::PageTimelineAgent::didScheduleStyleRecalculation):
(WebCore::PageTimelineAgent::willRecalculateStyle):
(WebCore::PageTimelineAgent::didRecalculateStyle):
(WebCore::PageTimelineAgent::willComposite):
(WebCore::PageTimelineAgent::didComposite):
(WebCore::PageTimelineAgent::willPaint):
(WebCore::PageTimelineAgent::didPaint):
(WebCore::PageTimelineAgent::mainFrameStartedLoading):
(WebCore::PageTimelineAgent::mainFrameNavigated):
(WebCore::PageTimelineAgent::didCompleteRenderingFrame):
(WebCore::PageTimelineAgent::shouldStartHeapInstrument const):
(WebCore::PageTimelineAgent::captureScreenshot):
* Source/WebCore/inspector/agents/worker/WorkerTimelineAgent.h: Added.
* Source/WebCore/inspector/agents/worker/WorkerTimelineAgent.cpp: Added.
(WebCore::WorkerTimelineAgent::WorkerTimelineAgent):
(WebCore::WorkerTimelineAgent::setAutoCaptureEnabled):
* Source/WebCore/inspector/InstrumentingAgents.h:
* Source/WebCore/inspector/InspectorController.cpp:
(WebCore::InspectorController::createLazyAgents):
* Source/WebCore/inspector/InspectorInstrumentation.cpp:
(WebCore::InspectorInstrumentation::didInvalidateLayoutImpl):
(WebCore::InspectorInstrumentation::willLayoutImpl):
(WebCore::InspectorInstrumentation::didLayoutImpl):
(WebCore::InspectorInstrumentation::willCompositeImpl):
(WebCore::InspectorInstrumentation::didCompositeImpl):
(WebCore::InspectorInstrumentation::willPaintImpl):
(WebCore::InspectorInstrumentation::didPaintImpl):
(WebCore::InspectorInstrumentation::willRecalculateStyleImpl):
(WebCore::InspectorInstrumentation::didRecalculateStyleImpl):
(WebCore::InspectorInstrumentation::didScheduleStyleRecalculationImpl):
(WebCore::InspectorInstrumentation::didCommitLoadImpl):
(WebCore::InspectorInstrumentation::frameStartedLoadingImpl):
(WebCore::InspectorInstrumentation::didCompleteRenderingFrameImpl):
Split the existing `InspectorTimelineAgent` into a `PageTimelineAgent` and `WorkerTimelineAgent`.
This is mainly done for clarity as certain concepts (e.g. compositing, navigation, screenshots, etc.) don't make sense in a `Worker`.

* Source/WebCore/workers/WorkerOrWorkletScriptController.h:
* Source/WebCore/workers/WorkerOrWorkletScriptController.cpp:
(WebCore::WorkerOrWorkletScriptController::evaluate):
(WebCore::WorkerOrWorkletScriptController::evaluateModule):
* Source/WebCore/bindings/js/ScriptModuleLoader.cpp:
(WebCore::ScriptModuleLoader::evaluate):
* Source/WebCore/inspector/InspectorInstrumentation.h:
(WebCore::InspectorInstrumentation::willEvaluateScript):
(WebCore::InspectorInstrumentation::didEvaluateScript):
Add `InspectorInstrumentation` hooks before and after evaluating JS in `Worker`.

* Source/JavaScriptCore/inspector/agents/InspectorHeapAgent.h:
* Source/WebCore/inspector/agents/WebHeapAgent.h:
* Source/WebCore/inspector/agents/WebHeapAgent.cpp:
(WebCore::WebHeapAgent::didCreateFrontendAndBackend):
(WebCore::WebHeapAgent::willDestroyFrontendAndBackend):
Also expose `WebHeapAgent` via `InstrumentingAgents` instead of only just `PageHeapAgent`.
This is needed by `InspectorTimelineAgent` to start capturing heap snapshots during a timeline recording.

* Source/WebInspectorUI/UserInterface/Models/Instrument.js:
(WI.Instrument.startLegacyTimelineAgent):
(WI.Instrument.stopLegacyTimelineAgent):
Iterate over all `WI.Target` when starting a timeline recording now that `Timeline` also exists in `Worker`.

* Source/WebInspectorUI/UserInterface/Protocol/TimelineObserver.js:
(WI.TimelineObserver.prototype.eventRecorded):
* Source/WebInspectorUI/UserInterface/Controllers/HeapManager.js:
(WI.HeapManager.prototype.garbageCollected):
Pass along the `WI.Target`.

* Source/WebInspectorUI/UserInterface/Controllers/TimelineManager.js:
(WI.TimelineManager):
(WI.TimelineManager.prototype.initializeTarget):
(WI.TimelineManager.prototype.capturingStarted):
(WI.TimelineManager.prototype.eventRecorded):
(WI.TimelineManager.prototype.scriptProfilerTrackingUpdated):
(WI.TimelineManager.prototype.scriptProfilerTrackingCompleted):
(WI.TimelineManager.prototype._processRecord):
(WI.TimelineManager.prototype._processEvent):
(WI.TimelineManager.prototype._stackTraceFromPayload):
(WI.TimelineManager.prototype._garbageCollected):
Have a separate array of `WI.ScriptTimelineRecord` for each `WI.Target`.

* Source/WebInspectorUI/UserInterface/Models/TimelineRecording.js:
(WI.TimelineRecording):
(WI.TimelineRecording.async import):
(WI.TimelineRecording.prototype.get targets): Added.
(WI.TimelineRecording.prototype.callingContextTree): Added.
(WI.TimelineRecording.prototype.reset):
(WI.TimelineRecording.prototype.updateCallingContextTrees): Renamed from `initializeCallingContextTrees`.
(WI.TimelineRecording.prototype.get topDownCallingContextTree): Deleted.
(WI.TimelineRecording.prototype.get bottomUpCallingContextTree): Deleted.
(WI.TimelineRecording.prototype.get topFunctionsTopDownCallingContextTree): Deleted.
(WI.TimelineRecording.prototype.get topFunctionsBottomUpCallingContextTree): Deleted.
Have separate `WI.CallingContextTree` for each `WI.Target` (and for each `WI.CallingContextTree.Type`).
Callers must `updateCallingContextTrees` with a valid `WI.Target` before attempting to fetch a `WI.CallingContextTree` for that `WI.Target` (and `WI.CallingContextTree.Type`).

* Source/WebInspectorUI/UserInterface/Models/ScriptTimelineRecord.js:
(WI.ScriptTimelineRecord.async fromJSON):
(WI.ScriptTimelineRecord.prototype.toJSON):
(WI.ScriptTimelineRecord.prototype.get target): Added.
(WI.ScriptTimelineRecord.prototype._initializeProfileFromPayload.profileNodeFromPayload):
* Source/WebInspectorUI/UserInterface/Models/CallingContextTree.js:
(WI.CallingContextTree):
(WI.CallingContextTree.prototype.get target): Added.
(WI.CallingContextTree.prototype.reset): Deleted.
(WI.CallingContextTree.__test_makeTreeFromProtocolMessageObject): Deleted.
(WI.CallingContextTree.prototype.__test_matchesStackTrace): Deleted.
(WI.CallingContextTree.prototype.__test_buildLeafLinkedLists): Deleted.
* Source/WebInspectorUI/UserInterface/Views/ProfileDataGridTree.js:
(WI.ProfileDataGridTree.prototype.get target): Added.
* Source/WebInspectorUI/UserInterface/Views/ProfileDataGridNode.js:
(WI.ProfileDataGridNode.prototype.iconClassName):
(WI.ProfileDataGridNode.prototype.filterableDataForColumn):
(WI.ProfileDataGridNode.prototype._displayContent):
Save and use the relevant `WI.Target`.

* Source/WebInspectorUI/UserInterface/Views/ScriptDetailsTimelineView.js:
(WI.ScriptDetailsTimelineView):
(WI.ScriptDetailsTimelineView.prototype.closed):
(WI.ScriptDetailsTimelineView.prototype._processPendingRecords):
* Source/WebInspectorUI/UserInterface/Views/ScriptTimelineDataGridNode.js:
(WI.ScriptTimelineDataGridNode.prototype.get data):
Show the name of the `WI.Target` (a.k.a. "Context") when there's more than one.

* Source/WebInspectorUI/UserInterface/Views/ScriptProfileTimelineView.js:
(WI.ScriptProfileTimelineView):
(WI.ScriptProfileTimelineView.prototype.closed):
(WI.ScriptProfileTimelineView.prototype.get navigationItems):
(WI.ScriptProfileTimelineView.prototype._callingContextTreeForOrientation):
(WI.ScriptProfileTimelineView.prototype._updateTargetNavigationItemDisplay):
(WI.ScriptProfileTimelineView.prototype._displayNameForTarget):
(WI.ScriptProfileTimelineView.prototype._profileOrientationButtonClicked):
(WI.ScriptProfileTimelineView.prototype._topFunctionsButtonClicked):
(WI.ScriptProfileTimelineView.prototype._createProfileView):
(WI.ScriptProfileTimelineView.prototype._showProfileView):
(WI.ScriptProfileTimelineView.prototype._handleRecordingTargetAdded):
(WI.ScriptProfileTimelineView.prototype._populateTargetNavigationItemContextMenu):
(WI.ScriptProfileTimelineView.prototype._showProfileViewForOrientation): Deleted.
* Source/WebInspectorUI/UserInterface/Views/ScriptProfileTimelineView.css: Added.
(.content-view.timeline-recording .navigation-bar > .script-profile-target > .selector-arrows):
Create a dropdown for all `WI.Target` when there's more than one to control which `WI.CallingContextTree` is shown.
We probably don't want to merge stack trace data from different `WI.Target` as it could be misleading if a function (or at least it's name and source code location) is used in multiple `WI.Target`.

* Source/WebInspectorUI/UserInterface/Views/ClusterContentView.js:
(WI.ClusterContentView):
(WI.ClusterContentView.prototype.closed):
(WI.ClusterContentView.prototype._contentViewNavigationItemsDidChange):
This appears to be the first time that a `WI.ContentView` inside a `WI.ClusterContentView` ever dynamically changes the result of `get navigationItems`, so dispatch a `WI.ContentView.Event.NavigationItemsDidChange` to the parent `WI.ContentBrowser` so that it's updated accordingly.

* Source/WebInspectorUI/UserInterface/Views/DataGrid.js:
(WI.DataGrid.prototype.insertColumn): All cells need to be refreshed when a new column is inserted.
(WI.DataGrid.prototype.setColumnVisible): When a new column is first inserted it doesn't have a previous `hidden` state, so always make sure to set one.
(WI.DataGrid.prototype._contextMenuInHeader): Columns should be listed by ordinal instead of insertion order.

* Source/WebCore/page/PerformanceUserTiming.cpp:
(WebCore::PerformanceUserTiming::mark):
* Source/WebInspectorUI/UserInterface/Models/ScriptInstrument.js:
(WI.ScriptInstrument.prototype.stopInstrumentation):
Drive-by: remove unused variable.

* Source/WebInspectorUI/UserInterface/Views/CPUTimelineView.js:
(WI.CPUTimelineView.prototype.closed):
* Source/WebInspectorUI/UserInterface/Views/HeapAllocationsTimelineView.js:
(WI.HeapAllocationsTimelineView.prototype.closed):
* Source/WebInspectorUI/UserInterface/Views/LayoutTimelineView.js:
(WI.LayoutTimelineView.prototype.closed):
* Source/WebInspectorUI/UserInterface/Views/MemoryTimelineView.js:
(WI.MemoryTimelineView.prototype.closed):
* Source/WebInspectorUI/UserInterface/Views/NetworkTimelineView.js:
(WI.NetworkTimelineView.prototype.closed):
* Source/WebInspectorUI/UserInterface/Views/OverviewTimelineView.js:
(WI.OverviewTimelineView.prototype.closed):
* Source/WebInspectorUI/UserInterface/Views/RenderingFrameTimelineView.js:
(WI.RenderingFrameTimelineView.prototype.closed):
Drive-by: add missing `super` call.

* Source/WebInspectorUI/UserInterface/Views/SettingsTabContentView.js:
(WI.SettingsTabContentView.prototype._createExperimentalSettingsView):
Drive-by: add missing separator.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebInspectorUI/UserInterface/Main.html:
* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:

* Source/WebInspectorUI/UserInterface/Models/CallingContextTreeNode.js:
(WI.CallingContextTreeNode.prototype.__test_buildLeafLinkedLists): Deleted.
* LayoutTests/inspector/sampling-profiler/basic.html:
* LayoutTests/inspector/sampling-profiler/call-frame-with-dom-functions.html:
* LayoutTests/inspector/sampling-profiler/eval-source-url.html:
* LayoutTests/inspector/sampling-profiler/expression-location-info.html:
* LayoutTests/inspector/sampling-profiler/expression-location-info-expected.txt:
* LayoutTests/inspector/sampling-profiler/many-call-frames.html:
* LayoutTests/inspector/sampling-profiler/named-function-expression.html:
* LayoutTests/inspector/sampling-profiler/resources/calling-context-tree.js: Added.
Move test functions to test files so they're not shipped in production builds.

Canonical link: https://commits.webkit.org/290430@main
